pkg/server/plugin/upstreamca/spire/spire.go

```diff
@@ -0,0 +1,184 @@
+package spireplugin
+
+import (
+	"context"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"sync"
+	//"time"
+
+	"github.com/hashicorp/hcl"
+
+	"github.com/spiffe/spire/pkg/common/x509svid"
+	"github.com/spiffe/spire/pkg/common/x509util"
+	"github.com/spiffe/spire/proto/server/upstreamca"
+	"github.com/spiffe/spire/api/workload"
+	"github.com/spiffe/spire/proto/common/plugin"
+
+	"crypto/ecdsa"
+	"github.com/spiffe/spire/pkg/common/bundleutil"
+	"google.golang.org/grpc"
+	"github.com/spiffe/go-spiffe/uri"
+	"github.com/spiffe/spire/pkg/common/idutil"
+	"net/url"
+)
+
+type Configuration struct {
+	TTL             string `hcl:"ttl" json:"ttl"` // time to live for generated certs
+	ServerAddr      string `hcl:"server_address" json:"server_address"`
+	ServerPort      string `hcl:"server_port" json:"server_port"`
+	ServerAgentAddr string `hcl:"server_agent_address" json:"server_agent_address"`
+}
+
+type spirePlugin struct {
+	serialNumber      x509util.SerialNumber
+	workloadApiClient workload.X509Client
+
+	mtx               sync.RWMutex
+	cert              *x509.Certificate
+	upstreamCA        *x509svid.UpstreamCA
+
+	workloadKey       []byte
+	workloadCert      []byte
+	workloadBundle    []byte
+
+	spiffeID          string
+	trustDomain       url.URL
+	privateKey        *ecdsa.PrivateKey
+	intermediateCA    []*x509.Certificate
+	bundle            *bundleutil.Bundle
+
+	conn              *grpc.ClientConn
```

Comment:
The UpstreamCA plugin is invoked only every now and then, when intermediate CA rotation is required (~1/2 CA TTL). The lifecycle of the invocation is also fairly short (CSR in, Certificate out). I think we can reduce a lot of the complexity in this plugin by not keeping workload api client, X509-SVID etc, cached on the struct here. Instead, when `SubmitCSR` is called, we can create a client, get the SVID, call the upstream server, and then throw it all away. The extra overhead is minimal

Reply:
Changes in SubmitCSR is changed to use local var instead of caching the connection are completed.